aten/src/ATen/OpaqueTensorImpl.h

```diff
@@ -99,6 +100,7 @@ c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach() const override {
   impl->is_contiguous_ = is_contiguous_;
   impl->is_wrapped_number_ = is_wrapped_number_;
   impl->reserved_ = reserved_;
+  impl->set_version_counter(version_counter);
```

Comment:
are there now calls to set_version_counter that aren't contained in `shallow_copy_and_detach`?

Reply:
There are still two of them:
1. https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/variable.cpp#L196. For this one, because `diff_view_meta->base_` can be set to `base` or `base.base()` depending on whether `base` is already a view, we don't know the right value for the view's version counter when we call `shallow_copy_and_detach` in https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/variable.h#L549, and we will have to wait until this step to set the version counter to the correct value.
2. https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/saved_variable.cpp#L87. For this one, since the `shallow_copy_and_detach` call is inside `make_variable()`, we can either allow passing `version_counter` as parameter to `make_variable()`, or treat this as a one-off case and allow calling `set_version_counter` outside of `shallow_copy_and_detach`.

Reply:
Ok, it sounds like both of these cases should go away soon:
1) We shouldn't need make_variable, because everything will be a variable.  And if we have a similar API -- we can just pass the version counter, as you mentioned.
2) I'm not sure why we can't just save the variable, instead of the tensor in the future -- maybe there is some infinite loop thing?

Reply:
1. Sounds great, yes we will be able to simplify this after we make everything a variable. 
2. I think currently we don't save the original variable because of https://bit.ly/2w367de (I wasn't able to use the Github link in this comment).

I added the task for looking into this in https://github.com/pytorch/pytorch/issues/13638.